openmmtools/integrators.py

```diff
@@ -1199,12 +1123,411 @@ def substep_function(step_string):
+        """
+        #convert the string to all caps
+        splitting_string = splitting_string.upper()
+
+        #sanity check the splitting string
+        self.sanity_check(splitting_string)
+
+        ORV_counts = dict()
+
+        #count number of R, V, O steps:
+        ORV_counts["R"] = splitting_string.count("R")
+        ORV_counts["V"] = splitting_string.count("V")
+        ORV_counts["O"] = splitting_string.count("O")
+
+        #split by delimiter (space)
+        step_list = splitting_string.split(" ")
+
+        #populate a list with all the force groups in the system
+        force_group_list = []
+        for step in step_list:
+            #if the length of the step is greater than one, it has a digit after it
+            if step[0] == "V" and len(step) > 1:
+                force_group_list.append(step[1:])
+
+        #Make a set to count distinct force groups
+        force_group_set = set(force_group_list)
+
+        #check if force group list cast to set is longer than one
+        #If it is, then multiple force groups are specified
+        if len(force_group_set) > 1:
+            mts = True
+        else:
+            mts = False
+
+
+        #If the integrator is MTS, count how many times the V steps appear for each
+        if mts:
+            force_group_n_V = {force_group: 0 for force_group in force_group_set}
+            for step in step_list:
+                if step[0] == "V":
+                    #ensure that there are no V-all steps if it's MTS
+                    assert len(step) > 1
+                    #extract the index of the force group from the step
+                    force_group_idx = step[1:]
+                    #increment the number of V calls for that force group
+                    force_group_n_V[force_group_idx] += 1
+        else:
+            force_group_n_V = {"0": ORV_counts["V"]}
+
+        return ORV_counts, mts, force_group_n_V
+
+    def metropolize(self):
+        """
+        Add a Metropolization (based on shadow work) step to the integrator. When Metropolization occurs, shadow work
+        is reset.
+        """
+        self.addComputeGlobal("accept", "step(exp(-(shadow_work)/kT) - uniform)")
+        self.beginIfBlock("accept != 1")
+        self.addComputePerDof("x", "xold")
+        self.addComputePerDof("v", "-vold")
+        self.addComputeGlobal("shadow_work", 0)
+        self.endBlock()
+
+    def begin_metropolize(self):
+        """
+        Save the current x and v for a metropolization step later
+        """
+        self.addComputePerDof("xold", "x")
+        self.addComputePerDof("vold", "v")
+
+class AlchemicalLangevinSplittingIntegrator(LangevinSplittingIntegrator):
+    """
+    This class performs nonequilibrium switching using the Perses-style lambda functions.
+
+    The type of propagator can be specified via the splitting string argument, described below.
+
+        Parameters
```

Comment:
Should the argument descriptions be in the docstring of the constructor?

Also, list of arguments in docstring is missing `direction`, `nsteps_neq`